src/burnchains/burnchain.rs

```diff
@@ -273,67 +306,215 @@ impl Burnchain {
+            let leader_key_vtxindex = block_candidates[i].key_vtxindex as u32;
+            let leader_key_opt = BurnDB::<A, K>::get_leader_key_at(tx, leader_key_block_height, leader_key_vtxindex)?;
+
+            match leader_key_opt {
+                None => {
+                    // should never happen; otherwise the commit would never have been accepted 
+                    panic!("No leader key for block commit {} (at {},{})", &block_candidates[i].txid.to_hex(), block_candidates[i].block_number, block_candidates[i].vtxindex);
+                },
+                Some(leader_key) => {
+                    leader_keys.push(leader_key)
+                }
+            };
+        }
+
+        Ok(leader_keys)
+    }
+
+    /// Append a block's checked transactions to the ledger and return the burn distribution
+    /// * insert all checked operations
+    /// * calculate a burn distribution
+    /// * return the burn distribution
+    fn append_blockstack_ops<'a, A, K>(tx: &mut Transaction<'a>, block_ops: &Vec<BlockstackOperationType<A, K>>) -> Result<Vec<BurnSamplePoint<A, K>>, burnchain_error>
+    where 
+        A: Address,
+        K: PublicKey
+    {
+        // block commits and support burns discovered in this block.
+        let mut block_commits: Vec<LeaderBlockCommitOp<A, K>> = vec![];
+        let mut user_burns: Vec<UserBurnSupportOp<A, K>> = vec![];
+
+        // store all leader VRF keys and block commits we found.
+        // don't store user burns until we know if they match a block commit.
+        for i in 0..block_ops.len() {
+            match block_ops[i] {
+                BlockstackOperationType::LeaderKeyRegister(ref op) => {
+                    Burnchain::store_transaction(tx, &block_ops[i])?;
+                },
+                BlockstackOperationType::LeaderBlockCommit(ref op) => {
+                    Burnchain::store_transaction(tx, &block_ops[i])?;
+                    block_commits.push(op.clone());
+                },
+                BlockstackOperationType::UserBurnSupport(ref op) => {
+                    user_burns.push(op.clone());
+                }
+            };
+        }
+
+        // find all VRF leader keys that were consumed by the leader block commits of this block 
+        let consumed_leader_keys_res = Burnchain::get_consumed_leader_keys(tx, &block_commits);
+        let consumed_leader_keys = consumed_leader_keys_res
+            .map_err(burnchain_error::DBError)?;
+
+        // calculate the burn distribution from these operations.
+        // The resulting distribution will contain the user burns that match block commits.
+        let burn_dist = BurnSamplePoint::make_distribution(block_commits, consumed_leader_keys, user_burns);
+        
+        // store user burns in the burn distribution -- these are the subset of user burns
+        // that matched a (previous) leader key and a (current) block commit.
+        for i in 0..burn_dist.len() {
```

Comment:
You can iterate here directly:

```
for burn_point in &burn_dist {
 ...
}
```

And similarly for a lot of other loops. This not only makes code more clear, it also lets rust omit range checks on the indexes, so its ever so slightly faster.

Reply:
Can do :+1: